tokio-net/src/driver/reactor/sharded_slab/slab.rs

```diff
@@ -0,0 +1,302 @@
+            shard.remove_local(idx)
+        } else {
+            shard.remove_remote(idx)
+        }
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the slab does not contain a value for the given key, `None` is
+    /// returned instead.
+    pub(in crate::driver) fn get(&self, token: usize) -> Option<&ScheduledIo> {
+        let tid = Tid::from_packed(token);
+        #[cfg(test)]
+        test_println!("get {:#x}; tid={:?}", token, tid);
+        self.shards.get(tid.as_usize())?.get(token)
+    }
+
+    /// Returns an iterator over all the items in the slab.
+    pub(in crate::driver::reactor) fn unique_iter(&mut self) -> iter::UniqueIter<'_> {
+        let mut shards = self.shards.iter_mut();
+        let shard = shards.next().expect("must be at least 1 shard");
+        let mut pages = shard.iter();
+        let slots = pages.next().and_then(page::Shared::iter);
+        iter::UniqueIter {
+            shards,
+            slots,
+            pages,
+        }
+    }
+}
+
+impl SingleShard {
+    /// Returns a new slab with the default configuration parameters.
+    pub(crate) fn new() -> Self {
+        Self {
+            shard: Shard::new(0),
+            local: Mutex::new(()),
+        }
+    }
+
+    /// allocs a value into the slab, returning a key that can be used to
+    /// access it.
+    ///
+    /// If this function returns `None`, then the shard for the current thread
+    /// is full and no items can be added until some are removed, or the maximum
+    /// number of shards has been reached.
+    pub(crate) fn alloc(&self) -> Option<usize> {
+        // we must lock the slab to alloc an item.
+        let _local = self.local.lock().unwrap();
+        #[cfg(test)]
+        test_println!("alloc");
+        self.shard.alloc()
+    }
+
+    /// Removes the value associated with the given key from the slab.
+    pub(crate) fn remove(&self, idx: usize) {
+        // try to lock the slab so that we can use `remove_local`.
+        let lock = self.local.try_lock();
+        // if we were able to lock the slab, we are "local" and can use the fast
+        // path; otherwise, we will use `remove_remote`.
+        let is_local = lock.is_ok();
+        test_println!("rm {:#x}; is_local={}", idx, is_local);
+
+        if is_local {
+            self.shard.remove_local(idx)
+        } else {
+            self.shard.remove_remote(idx)
+        }
+    }
+
+    /// Return a reference to the value associated with the given key.
+    ///
+    /// If the slab does not contain a value for the given key, `None` is
+    /// returned instead.
+    pub(in crate::driver) fn get(&self, token: usize) -> Option<&ScheduledIo> {
+        #[cfg(test)]
+        test_println!("get {:#x}", token);
+        self.shard.get(token)
+    }
+
+    /// Returns an iterator over all the items in the slab.
+    pub(in crate::driver::reactor) fn unique_iter(&self) -> iter::ShardIter<'_> {
```

Comment:
Pretty minor, but i think you can still take `&mut self` here and avoid the lock?

Reply:
yup, overlooked that, thanks.